pkg/platform/runtime/artifactvalidator/validator.go

```diff
@@ -0,0 +1,152 @@
+package artifactvalidator
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"strings"
+
+	"github.com/ActiveState/cli/internal/download"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/fileutils"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"go.mozilla.org/pkcs7"
+)
+
+type signature struct {
+	Sig  string `json: "sig"`
+	Cert string `json: "cert"`
+}
+
+type attestation struct {
+	Payload    string      `json: "payload"`
+	Signatures []signature `json: "signatures"`
+}
+
+func ValidateAttestation(attestationFile string) error {
+	data, err := fileutils.ReadFile(attestationFile)
+	if err != nil {
+		return errs.Wrap(err, "Could not read attestation: "+attestationFile)
+	}
+
+	att := attestation{}
+	err = json.Unmarshal(data, &att)
+	if err != nil {
+		return errs.Wrap(err, "Could not unmarshal attestation")
+	}
+
+	if len(att.Signatures) == 0 {
+		return locale.NewError("validate_attestation_fail_no_signatures", "No signatures")
+	}
+
+	// Verify signing certificate.
+	pemBlock, _ := pem.Decode([]byte(att.Signatures[0].Cert))
+	if pemBlock == nil {
+		return errs.Wrap(err, "Unable to decode attestation certificate")
+	}
+
+	cert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return errs.Wrap(err, "Unable to parse attestation certificate")
+	}
+
+	intermediates := x509.NewCertPool()
+	addIntermediatesToPool(cert, intermediates)
+
+	opts := x509.VerifyOptions{
+		Roots:         nil, // use system root CAs
+		Intermediates: intermediates,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning},
+	}
+	_, err = cert.Verify(opts)
+	if err != nil {
+		return errs.Wrap(err, "Unable to validate certificate")
+	}
+
+	// Verify signature.
+	payload := make([]byte, len(att.Payload))
+	n, err := base64.StdEncoding.Decode(payload, []byte(att.Payload))
+	if err != nil {
+		return errs.Wrap(err, "Unable to decode attestation payload")
+	}
+	payload = payload[:n]
+	hash := sha256.New()
+	hash.Write(payload)
+	digest := hash.Sum(nil)
+
+	signature := make([]byte, len(att.Signatures[0].Sig))
+	n, err = base64.StdEncoding.Decode(signature, []byte(att.Signatures[0].Sig))
+	if err != nil {
+		return errs.Wrap(err, "Unable to decode attestation signature")
+	}
+	signature = signature[:n]
+
+	publicKey, ok := cert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		return locale.NewError("validate_attestation_fail_public_key", "Certificate's public key is not an expected RSA pubkey")
+	}
+	err = rsa.VerifyPSS(publicKey, crypto.SHA256, digest, signature, &rsa.PSSOptions{Hash: crypto.SHA256})
+	if err != nil {
+		return errs.Wrap(err, "Unable to validate signature")
+	}
+
+	// TODO: read payload artifact SHAs and validate them against downloaded artifact SHAs.
```

Comment:
Note that you will likely end up verifying checksums on artifacts you wouldn't otherwise download.  The attestations will contain checksums for the entire build closure of artifacts some of which may not be in the runtime closure of the build e.g. the builders and their dependences and build tools.

Reply:
I didn't exactly understand what the attestation payload contained, but the idea I had in mind was for all artifacts we downloaded from the build, look them up in the attestation payload and see if there's a checksum. If there is, we validate against that. Otherwise, fall back on the checksum provided by the Platform outside the attestation. I was not going to iterate over payload checksums and validate them, because you're right, they contain checksums for artifacts that wouldn't be in the runtime.